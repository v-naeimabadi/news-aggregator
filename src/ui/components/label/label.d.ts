import { DetailedHTMLProps, LabelHTMLAttributes, ReactElement } from "react";
import { FieldPath, FieldValues } from "react-hook-form";
import { StyleProp, TextProps, ViewStyle } from "react-native";
import { CSSProperties } from "styled-components";

export interface LabelPropsBase<
  TFieldValues extends FieldValues = FieldValues,
  TFieldName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
> {
  required?: boolean;
  color?: string;
  error?: boolean;
  connect?: TFieldName | string;
}

export interface LabelPropsWeb<
  TFieldValues extends FieldValues = FieldValues,
  TFieldName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
> extends LabelPropsBase<TFieldValues, TFieldName>,
    Omit<
      DetailedHTMLProps<
        LabelHTMLAttributes<HTMLLabelElement>,
        HTMLLabelElement
      >,
      "ref"
    > {
  containerStyle?: CSSProperties;
  ref?: RefObjectt<HTMLLabelElement> | null | undefined;
}

export interface LabelPropsNative<
  TFieldValues extends FieldValues = FieldValues,
  TFieldName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
> extends LabelPropsBase<TFieldValues, TFieldName>,
    TextProps {
  containerStyle?: StyleProp<ViewStyle>;
}

export type LabelProps<
  TFieldValues extends FieldValues = FieldValues,
  TFieldName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>,
> =
  | LabelPropsWeb<TFieldValues, TFieldName>
  | LabelPropsNative<TFieldValues, TFieldName>;

export function Label(props: LabelProps<never, never>): ReactElement | null;
