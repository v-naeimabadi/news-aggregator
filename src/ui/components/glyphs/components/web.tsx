import { forwardRef, ReactElement, SVGProps, LegacyRef } from "react";
import styled from "styled-components";

function WebBase(
  props: SVGProps<SVGSVGElement>,
  ref: LegacyRef<SVGSVGElement>,
): ReactElement {
  return (
    <svg width="26px" height="26px" viewBox="0 0 26 26" ref={ref} {...props}>
      <g>
        <g transform="translate(-388, -204)">
          <g transform="translate(390, 206)">
            <path d="M11,0 C17.0581597,0 22,4.94982474 22,10.9999395 C22,13.6682058 21.022889,16.2083146 19.3027869,18.200118 L19.1674654,18.3508696 C19.1512377,18.3912237 19.1290406,18.4266334 19.0993436,18.4634929 L19.0489817,18.5206686 L18.9862464,18.5847233 C16.9080459,20.7414998 14.0386281,22 11,22 C4.93494566,22 0,17.0829962 0,11.0397943 C0,4.98262691 4.94886312,0 11,0 Z M12.0394262,16.9227826 L12.0394262,19.7378261 L12.1133248,19.7033936 C12.8608534,19.3167781 13.5616933,18.5405755 14.1303348,17.4546617 L14.2426383,17.2307826 L13.89331,17.158401 C13.4353858,17.0709643 12.97022,17.0047775 12.4978445,16.9598707 L12.0394262,16.9227826 Z M9.99941498,16.9227826 L9.64464682,16.9499585 C9.08195814,16.9989129 8.53191542,17.0768251 8.00839197,17.1821425 L7.80484292,17.226 L7.92137675,17.4494445 C8.50987119,18.5388787 9.21354138,19.317023 9.94633561,19.7040325 L9.99941498,19.7301739 L9.99941498,16.9227826 Z M5.79075184,17.8706957 L5.6598675,17.9250494 L5.40482972,18.0371304 L5.46324285,18.084971 C5.66157744,18.241542 5.86743948,18.3893304 6.08054085,18.5282104 L6.15747386,18.5756522 L6.08547363,18.4499385 C6.02676884,18.3431193 5.96948867,18.2345565 5.9137081,18.1243426 L5.79075184,17.8706957 Z M16.178009,17.8783478 L16.0687512,18.1191512 C16.0148316,18.2308022 15.9589208,18.3407054 15.9010927,18.4487694 L15.8439271,18.5498261 L15.8790372,18.5280378 C16.0921069,18.3891737 16.2979272,18.2413966 16.496221,18.0848247 L16.546651,18.041913 L16.3393997,17.9481412 L16.178009,17.8783478 Z M19.9172295,12.035913 L17.5421765,12.035913 L17.5383698,12.1291966 C17.4683097,13.427946 17.2616899,14.6724565 16.9285007,15.8593642 L16.9018529,15.9471304 L17.1244286,16.0448197 C17.4173042,16.176213 17.6860796,16.3067887 17.9435101,16.4435947 L18.0864994,16.522 L18.2643535,16.2942034 C19.1402616,15.1147548 19.6978035,13.7359037 19.895341,12.2334916 L19.9172295,12.035913 Z M4.45730451,12.035913 L2.08417146,12.035913 L2.10517445,12.2149855 C2.30544311,13.6870576 2.87607867,15.0901621 3.74512109,16.2905081 L3.91106151,16.5105217 L4.16905054,16.366551 C4.38491278,16.2507775 4.6065526,16.1425128 4.83230782,16.0427693 L5.091868,15.9327826 L5.0858617,15.910136 C4.78260244,14.8431862 4.58181589,13.7295745 4.49089891,12.572048 L4.45730451,12.035913 Z M15.4983252,12.035913 L12.0394262,12.035913 L12.0394262,14.8853913 L12.4844308,14.9148818 C13.2306987,14.9719212 13.9465379,15.0704364 14.6202665,15.2115675 L14.9501622,15.2852174 L14.9539364,15.2751946 C15.2052181,14.3786806 15.37979,13.4413613 15.4650809,12.4946377 L15.4983252,12.035913 Z M9.99941498,12.035913 L6.54051596,12.035913 L6.5751205,12.4950384 C6.66040737,13.4420262 6.83496177,14.3793536 7.08623777,15.275424 L7.09829196,15.2807749 C7.8779516,15.0909374 8.67341367,14.9582697 9.48454644,14.8853005 L9.99941498,14.8471304 L9.99941498,12.035913 Z M3.91682154,5.55930435 L3.74171297,5.79248503 C2.86729095,6.99567402 2.30263875,8.37899191 2.10448464,9.85874936 L2.08321145,10.0425217 L4.45730451,10.0425217 L4.46168731,9.95063405 C4.53174743,8.65188468 4.73836725,7.40737419 5.07155644,6.2204665 L5.09666802,6.13226087 L4.8756285,6.03501102 C4.58275293,5.90361768 4.31397756,5.77304201 4.05654705,5.63623594 L3.91682154,5.55930435 Z M18.0903395,5.52773913 L17.8309352,5.67349603 C17.6150729,5.78926945 17.3934331,5.89753417 17.1676779,5.99727773 L16.906653,6.10643478 L16.914124,6.12991099 C17.260706,7.34928223 17.4734461,8.62959949 17.5413029,9.96676249 L17.5421765,10.0033043 L19.9162695,10.0033043 L19.8952479,9.81899267 C19.6976945,8.3390502 19.1396521,6.95935614 18.2614495,5.75459219 L18.0903395,5.52773913 Z M14.9511222,6.75686957 L14.9416223,6.75941441 C14.1619627,6.94925199 13.3665006,7.08191965 12.5553678,7.15488885 L12.0394262,7.19208696 L12.0394262,10.0033043 L15.4983252,10.0033043 L15.4647938,9.54515095 C15.3937214,8.75599448 15.2606577,7.97354655 15.072863,7.21609236 L14.9511222,6.75686957 Z M7.04835876,6.754 L7.04604933,6.76499477 C6.79476765,7.66150873 6.6201957,8.59882804 6.53490479,9.54555161 L6.50019574,10.0033043 L9.99941498,10.0033043 L9.99941498,7.19208696 L9.97566624,7.19171104 C9.16197223,7.14312177 8.36029471,7.03483605 7.55873789,6.86801474 L7.04835876,6.754 Z M12.0394262,2.34921739 L12.0394262,5.1566087 L12.4982493,5.12013871 C12.9707623,5.07518533 13.4358932,5.00901236 13.8936909,4.92162505 L14.2330383,4.85052174 L14.1185375,4.63074202 C13.5300431,3.54130783 12.8263729,2.76316352 12.0935787,2.37615403 L12.0394262,2.34921739 Z M9.99941498,2.34156522 L9.92658943,2.37679289 C9.17906089,2.76340846 8.47822102,3.53961103 7.90957949,4.62552482 L7.79620287,4.8486087 L8.14660427,4.92178553 C8.60452843,5.00922218 9.06969432,5.07540904 9.54206979,5.12031579 L9.99941498,5.1566087 L9.99941498,2.34156522 Z M15.818007,3.51426087 L15.8568961,3.5820431 C15.9160266,3.69106562 15.9741049,3.80276821 16.0311746,3.9171726 L16.1712889,4.20965217 L16.3083234,4.15344983 L16.551451,4.04130435 L16.4964572,3.99514436 C16.2981226,3.83857338 16.0922605,3.690785 15.8791592,3.55190496 L15.818007,3.51426087 Z M6.0984556,3.63127507 L6.15459384,3.52956522 L6.12051994,3.5517929 C5.90745027,3.69065696 5.70162998,3.8384341 5.50333618,3.995006 L5.45186998,4.03652174 L5.66015748,4.13169721 L5.820512,4.20008696 L5.93080557,3.96084191 C5.98472248,3.84921134 6.04063044,3.73932537 6.0984556,3.63127507 Z"></path>
          </g>
        </g>
      </g>
    </svg>
  );
}

const RefForwardingWeb = forwardRef(WebBase);

export const Web = styled(RefForwardingWeb).withConfig({
  shouldForwardProp: (prop, defaultValidatorFn) =>
    !["width", "height", "color"].includes(prop) && defaultValidatorFn(prop),
})<{
  height: number;
  width: number;
  color: string;
}>`
  height: ${({ height }) => `${height}px`};
  width: ${({ width }) => `${width}px`};
  & > * {
    fill: ${({ color }) => color} !important;
  }
`;
