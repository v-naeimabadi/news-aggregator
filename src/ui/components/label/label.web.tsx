import { FunctionComponent } from "react";
import { useFormState } from "react-hook-form";
import styled from "styled-components";
import { LabelPropsWeb } from "./label";

function ConnectedLabel({
  connect,
  error,
  ...props
}: Omit<LabelPropsWeb, "connect"> & Required<Pick<LabelPropsWeb, "connect">>) {
  const { errors } = useFormState({ name: connect, exact: true });
  return <LabelBase error={error || !!errors[connect]} {...props} />;
}

function LabelBase({
  required,
  children,
  color,
  error,
  containerStyle,
  ...props
}: Omit<LabelPropsWeb, "connect" | "ref">) {
  return (
    <Container style={containerStyle}>
      <StyledLabel color={error ? "#FF3232" : color} {...props}>
        {children}
      </StyledLabel>
      {required ? (
        <RequiredIcon data-testid="label-required-icon">*</RequiredIcon>
      ) : null}
    </Container>
  );
}

export const Label: FunctionComponent<LabelPropsWeb> = ({
  connect,
  ...props
}) => {
  if (connect) return <ConnectedLabel connect={connect} {...props} />;
  return <LabelBase {...props} />;
};

const StyledLabel = styled("label")<{ color?: string }>`
  color: ${({ color }) => color || "#3c3c3c"};
  font-family: "Avenir Next", -apple-system, BlinkMacSystemFont, "Segoe UI",
    "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans",
    "Helvetica Neue", sans-serif;
  font-size: 14px;
  font-weight: 600;
  letter-spacing: 0px;
  line-height: 19px;
  margin-bottom: 6px;
  margin-right: 6px;
  text-align: left;
`;

const RequiredIcon = styled("span")`
  color: #ff3232;
  font-family: "Avenir Next", -apple-system, BlinkMacSystemFont, "Segoe UI",
    "Roboto", "Oxygen", "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans",
    "Helvetica Neue", sans-serif;
  font-size: 20px;
  font-weight: 700;
  margin-left: -1px;
  margin-top: -1px;
`;

const Container = styled("div")`
  display: flex;
  height: fit-content;
  justify-content: flex-start;
`;
