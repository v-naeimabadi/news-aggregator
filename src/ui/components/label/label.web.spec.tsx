import { render } from "@testing-library/react";
import { Label } from "./label.web";

describe("Label", () => {
  it("renders correctly", () => {
    expect(render(<Label>Title</Label>)).toMatchSnapshot();
  });

  it("renders children correctly", () => {
    const children = "Title";
    const { getByText } = render(<Label>{children}</Label>);
    expect(getByText(children)).toBeDefined();
  });

  test("required icon is shown", () => {
    const { getByTestId } = render(<Label required>Title</Label>);
    expect(getByTestId("label-required-icon")).toBeDefined();
  });
});
