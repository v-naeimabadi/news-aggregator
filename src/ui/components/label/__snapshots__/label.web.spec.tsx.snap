// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c1 {
  color: #3c3c3c;
  font-family: "Avenir Next",-apple-system,BlinkMacSystemFont,"Segoe UI", "Roboto","Oxygen","Ubuntu","Cantarell","Fira Sans","Droid Sans", "Helvetica Neue",sans-serif;
  font-size: 14px;
  font-weight: 600;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  line-height: 19px;
  margin-bottom: 6px;
  margin-right: 6px;
  text-align: left;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

<body>
    <div>
      <div
        class="c0"
      >
        <label
          class="c1"
        >
          Title
        </label>
      </div>
    </div>
  </body>,
  "container": .c1 {
  color: #3c3c3c;
  font-family: "Avenir Next",-apple-system,BlinkMacSystemFont,"Segoe UI", "Roboto","Oxygen","Ubuntu","Cantarell","Fira Sans","Droid Sans", "Helvetica Neue",sans-serif;
  font-size: 14px;
  font-weight: 600;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
  line-height: 19px;
  margin-bottom: 6px;
  margin-right: 6px;
  text-align: left;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
}

<div>
    <div
      class="c0"
    >
      <label
        class="c1"
      >
        Title
      </label>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
